sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeRow.java

```diff
@@ -0,0 +1,409 @@
+import org.apache.spark.sql.types.DataType;
+import static org.apache.spark.sql.types.DataTypes.*;
+
+import org.apache.spark.sql.types.StructField;
+import org.apache.spark.sql.types.StructType;
+import org.apache.spark.sql.types.UTF8String;
+import org.apache.spark.unsafe.PlatformDependent;
+import org.apache.spark.unsafe.bitset.BitSetMethods;
+import org.apache.spark.unsafe.string.UTF8StringMethods;
+import scala.collection.Map;
+import scala.collection.Seq;
+import scala.collection.mutable.ArraySeq;
+
+import javax.annotation.Nullable;
+import java.math.BigDecimal;
+import java.sql.Date;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+
+// TODO: pick a better name for this class, since this is potentially confusing.
+// Maybe call it UnsafeMutableRow?
+
+/**
+ * An Unsafe implementation of Row which is backed by raw memory instead of Java objets.
+ *
+ * Each tuple has three parts: [null bit set] [values] [variable length portion]
+ *
+ * The bit set is used for null tracking and is aligned to 8-byte word boundaries.  It stores
+ * one bit per field.
+ *
+ * In the `values` region, we store one 8-byte word per field. For fields that hold fixed-length
+ * primitive types, such as long, double, or int, we store the value directly in the word. For
+ * fields with non-primitive or variable-length values, we store a relative offset (w.r.t. the
+ * base address of the row) that points to the beginning of the variable-length field.
+ */
+public final class UnsafeRow implements MutableRow {
+
+  private Object baseObject;
+  private long baseOffset;
+  private int numFields;
+  /** The width of the null tracking bit set, in bytes */
+  private int bitSetWidthInBytes;
+  @Nullable
+  private StructType schema;
+
+  private long getFieldOffset(int ordinal) {
+   return baseOffset + bitSetWidthInBytes + ordinal * 8;
+  }
+
+  public static int calculateBitSetWidthInBytes(int numFields) {
+    return ((numFields / 64) + ((numFields % 64 == 0 ? 0 : 1))) * 8;
+  }
+
+  /**
+   * Field types that can be updated in place in UnsafeRows (e.g. we support set() for these types)
+   */
+  public static final Set<DataType> settableFieldTypes;
+
+  /**
+   * Fields types can be read(but not set (e.g. set() will throw UnsupportedOperationException).
+   */
+  public static final Set<DataType> readableFieldTypes;
+
+  static {
+    settableFieldTypes = new HashSet<DataType>(Arrays.asList(new DataType[] {
+      IntegerType,
+      LongType,
+      DoubleType,
+      BooleanType,
+      ShortType,
+      ByteType,
+      FloatType
+    }));
+
+    // We support get() on a superset of the types for which we support set():
+    readableFieldTypes = new HashSet<DataType>(Arrays.asList(new DataType[] {
+      StringType
+    }));
+    readableFieldTypes.addAll(settableFieldTypes);
+  }
+
+  public UnsafeRow() { }
+
+  public void set(Object baseObject, long baseOffset, int numFields, StructType schema) {
```

Comment:
i'd name this something else other than set, otherwise it sounds like we are updating some field. We are really updating everything here.